### PR TITLE
Improve incremental build times of the libraries subset

### LINKED ISSUE #49800: Most of libs partition is not incrementally buildable
Running `build clr+libs -ninja` and then `build libs -ninja` results in 217 csc invocations when there should be 0.

It looks like system which creates `*.Forwards.cs` files may not be incremental.

### PR DESCRIPTION
**Pull request by ViktorHofer**
Fixes https://github.com/dotnet/runtime/issues/49800, https://github.com/dotnet/runtime/issues/43118, https://github.com/dotnet/runtime/issues/47155, https://github.com/dotnet/runtime/issues/64616

**TL;DR: This change makes incrementally building the libs subset ~60% faster. 63s on my machine (Surface Book 2), 38s on @ericstj's machine.**

I'm still on holiday until April this year but I had some time and was kind of bored so I played around with the repo in its current state. I already knew from before I left that some parts of the libs subset aren't incrementally buildable and I wanted to understand why. By looking at a binlog which I generated by building the libraries subset twice (`build.cmd libs && build.cmd libs /bl`) I spotted multiple csc task invocations which indicate that projects rebuild as the compiler shouldn't be invoked in an incremental build.

The reasons for that are:
- [The shim assemblies always rebuild](https://github.com/dotnet/runtime/blob/eeaee44e6fe4425db044967e025255e6a12c97a9/src/libraries/ref.proj#L82-L89). That is by design because of the sequencing of inbox and out-of-box reference assemblies. (1)
- Out-of-band libraries implicitly reference the .NET Framework shims which cause them to rebuild as well.
- Test projects reference the `Microsoft.NET.Test.Sdk` nuget package (by design) which breaks incremental builds. I filed https://github.com/microsoft/vstest/issues/3266 so that the VSTest team can look into that.

(1) For a better understanding, here's a quick overview of how shim assemblies are built. There are two types of shim assemblies, the .NETStandard shim `netstandard.dll` and the .NETFramework shim assemblies, i.e. `mscorlib.dll`, `System.dll`, etc. The former is satisfied by the `Microsoft.NETCore.App` shared framework closure or to rephrase, all types exposed by the `netstandard.dll` shim are present in the `Microsoft.NETCore.App` shared framework assemblies. Whereas the .NETFramework shims aren't satisified by the NETCoreApp closure and require the out-of-band assemblies as well. Therefore to build the .NETFramework shims, both inbox and out-of-box assemblies are required to be built. That alone wouldn't be a problem per-se. The issue is that both shim types are part of the `Microsoft.NETCore.App` shared framework and hence to compose the entire shared framework including the shims, oob assemblies need to be available. To summarize, the `netstandard.dll` shim can be built along with the shared framework but the .NETFramework shims can't.

## Changes
I was able to mitigate the above issues by doing the following:

**Commit 1: Separate the ApiCompat and shims logic**
To make the subsequent changes easier, I moved the ApiCompat related files into their own directory away from `src/libraries/shims` to `src/libraries/apicompat`. This is clean-up step which makes the following changes easier.

<br />

**Commit 2: Shared framework and out-of-band project split, shim projects refactoring and leveraging ProjectReferences**

- Split the build of shared framework and out-of-band projects so that it's possible to just build the shard framework projects, which was requested by the CLR teams (cc @jkotas). If only reference assemblies of one of these subsets should be built, the `/p:RefOnly=true` switch can be passed in. Generate the targeting pack with an incomplete (without shims) frameworklist as part of the "libs.sfx" subset.
- Build the shim projects after the shared framework and oob projects and re-generate the targeting pack's frameworklist to include the shims.
- Refactor the shims so that they aren't grouped anymore by generated & manual. Instead group them by "ref" and "src" same as other libraries in the tree which allows to reference the source project shims and read from the `ReferenceAssembly` metadata.
- Use ProjectReferences in traversal projects and shim projects instead of named references that point to binplace directories (see graph below). This allows to build parts of the product dynamically, i.e. the shim projects, apicompat.proj, sfx.proj and oob.proj. The outcome of that is the following design:

```mermaid
graph TD
    subgraph libs
    A[Build.proj]
    end
    subgraph sfx [libs.sfx]
        A --> B[sfx.proj] --> C[sfx-src.proj]
        B --> M[sfx-ref.proj]
    end
    subgraph oob [libs.oob]
        A --> D[oob.proj] --> L[apicompat.proj] --> C & E & G
        D --> E[oob-src.proj]
        D --> F[oob-all.proj]
        D --> G[shims.proj] --> J[Example: mscorlib.csproj] --> K[netstandard.csproj] --> C
        G --> K
        J --> C & E
    end
```

**Commit 3: Use TargetFramework singular for test projects**
Use TargetFramework instead of TargetFrameworks property whenever a
test project targets a single target framework only.
This avoid unnecessary outer builds and evaluations and makes the build
faster.

## Results

After these optimizations, the msbuild log viewer shows these remaining most expensive tasks when incrementally building the the libs subset:

image

These numbers are from a Surface Book 2 when building with the newest SDK which includes various msbuild up-to-date file check optimizations: `7.0.100-preview.2.22111.4`.

### CODE REVIEW

**Comment by ViktorHofer** [comment]
@ericstj @safern would you mind checking out the following branch which contains these changes + https://github.com/dotnet/runtime/pull/64500 + a commit on top which converts all applicable source and ref projects from `TargetFrameworks` to `TargetFramework` + an SDK upgrade which builds faster:

```cmd
git add ViktorHofer https://github.com/ViktorHofer/runtime
git checkout LibsIntermentalFastSuperFastSafeBranch
build.cmd libs && build.cmd libs
```

Can you please share the time the incremental build takes when running on your machine? That would be really cool. I expect the incremental build to be **much** faster but I would like to have some more data.

**Comment by ericstj** [comment]
The time I get on `main` is `Time Elapsed 00:01:27.95`
On your branch I get `Time Elapsed 00:00:39.87`

Definitely an improvement.  I'll take a look at the binlog and see if I see any other opportunities for improvement.

**Comment by ericstj** [comment]
Binlog demonstrates that all the rebuild problems have been solved.  No more accidental rebuilds 🎉  
So at this point it's just down to optimizing the no-op path.

Here's a some of the most expensive targets, omitting any targets that use CallTarget or MSBuild.  This was from a command that took `00:00:38.92`  I gathered these with /clp:PerformanceSummary.  You can confirm some of the large durations in binlogs using $target Name $time to see the duration.
```
      499 ms  GenerateNETCompatibleDefineConstants     487 calls
      594 ms  CopyFilesToOutputDirectory               487 calls
      604 ms  CoreResGen                               481 calls
      612 ms  GetCopyToOutputDirectoryItems            487 calls
      659 ms  ResolvePackageAssets                     491 calls
      712 ms  _CleanGetCurrentAndPriorFileWrites       487 calls
      767 ms  GeneratePartialFacadeSource              101 calls
      842 ms  ResolveTargetingPackAssets               161 calls
      853 ms  _HandlePackageFileConflicts              487 calls
      949 ms  GetBinPlaceTargetFramework               474 calls
     1345 ms  BinPlaceFiles                            1218 calls
     1405 ms  FindReferenceAssembliesForReferences     487 calls
     2150 ms  CoreCompile                              487 calls
     3509 ms  BuildNativeWindows                         1 calls
     5372 ms  FilterImplicitAssemblyReferences         362 calls
     9433 ms  Restore                                    2 calls
     9709 ms  ResolveAssemblyReferences                487 call
```

RAR is pretty darn expensive.  Is it using the cache file?  Can we reduce that time at all?
FilterImplicitAssemblyReferences  seems *much* too expensive for what it does.  There should be some optimization we can to here.
FindReferenceAssembliesForReferences is also too expensive, though not quite as bad.
BinPlaceFiles is relying on SkipUnchangedFiles for incrementality, we could avoid this overhead if we just used inputs/outputs.

After this it may be getting into diminishing returns.

**Comment by ViktorHofer** [comment]
> Here's a some of the most expensive targets, omitting any targets that use CallTarget or MSBuild. This was from a command that took 00:00:38.92 I gathered these with /clp:PerformanceSummary. You can confirm some of the large durations in binlogs using $target Name $time to see the duration.

In addition to `/clp:PerformanceSummery` you can also leverage the `/profileevaluation` switch which embeds evaluation time into the binlog. That switch was added four years ago, see https://github.com/dotnet/msbuild/blob/main/documentation/evaluation-profiling.md for more details.

> So at this point it's just down to optimizing the no-op path.

From the data that I gathered, 55% of the fully incremental build time is spent on project evaluation. Note that project evaluation also dominates the time spent during the the Restore task (which evaluates all projects as well, but slightly different via static graph).

> FilterImplicitAssemblyReferences seems much too expensive for what it does. There should be some optimization we can to here.

YESSS. I knew about that target being quite expensive as it builds the cartesian product of references and excludes instead of just subtracting excludes from references. I will probably send a PR to fix.

> RAR is pretty darn expensive. Is it using the cache file? Can we reduce that time at all?

Unsure about the cache file, I expect it's in use but haven't checked. Overall RAR improvements are tracked via https://github.com/dotnet/msbuild/issues/2015#issuecomment-1029882319 in which I just recently provided some data from dotnet/runtime.

> FindReferenceAssembliesForReferences is also too expensive, though not quite as bad.

I believe the high amount of time spent on that target is because of batching on a large number of items/references: https://github.com/dotnet/msbuild/blob/de1d7a295d709c3a748cc46cd5dc9bddd624ea46/src/Tasks/Microsoft.Common.CurrentVersion.targets#L2379-L2396.

Unsure if batching could be avoided entirely here but at least some usages could be replaced with item functions `i.e. ->Metadata`.

**Comment by ViktorHofer** [comment]
@ericstj @safern this is the last of my changes that I planned to deliver and should be ready now. Please review by commit.

I opened PRs for few unrelated changes:
- https://github.com/dotnet/runtime/pull/65378
- https://github.com/dotnet/runtime/pull/65381
- https://github.com/microsoft/MSBuildSdks/pull/338

**Comment by joperezr** [comment]
Before actually looking at the changes here, I had a question about the following comment in your description:

>In the next infrastructure rollout, the current "libs.ref+libs.src" subsets should be removed and "libs.sfx+libs.oob" subsets should be added. Generate the targeting pack with an incomplete (without shims) frameworklist as part of the "libs.sfx" subset

I understand the need to separate the libs.src into what is in and what is out of the framework. Regarding the refs though, I think a bit differently, here are my main two points:
- Today, building libs.ref (which includdes OOBs) is SUPER fast, compared to libs.src. I think that most times I do it it finishes in under a minute in my machine. This point is mainly to illustrate that perhaps libs.ref shouldn't be split to sfx+oobs
- Building libs.ref is really handy in the sense that today, if you have a clean enlistment and want to be able to open a specific library project in VS quickly, you can do so by just running `build.cmd -restore -s libs` and `build.cmd -s libs.ref` both of which will take less than a minute. After that, you can then open a specific library (via `build.cmd -vs System.Text.RegularExpressions`) and everything should just work(intellisense, build, etc) except for running tests of course. If I'm understanding your description correctly, you plan on removing the libs.ref, so this argument is basically challenging that and asking to please keep that subset around as it is very useful for quickly opening a project and  having everything required to build the project ready.

**Comment by ViktorHofer** [comment]
@joperezr thanks for the valuable feedback. I didn't know that people use the libs.ref subset that way.

> Building libs.ref is really handy in the sense that today, if you have a clean enlistment and want to be able to open a specific library project in VS quickly, you can do so by just running build.cmd -restore -s libs and build.cmd -s libs.ref both of which will take less than a minute. After that, you can then open a specific library (via build.cmd -vs System.Text.RegularExpressions) and everything should just work(intellisense, build, etc) except for running tests of course. If I'm understanding your description correctly, you plan on removing the libs.ref, so this argument is basically challenging that and asking to please keep that subset around as it is very useful for quickly opening a project and having everything required to build the project ready.

I agree with your sentiment. I believe there are two independent desires:
1. Being able to enter VS as quickly as possible from a fresh clone and
2. Being able to build just the reference assemblies (for whatever reason).

Both desires are legit and not dotnet/runtime specific at all as support for both exist in the product. Being able to build from within VS without any upfront steps is possible via using `ProjectReference` items to declare dependencies (instead of i.e. using loosely couple mechanisms like `Reference` items). The out-of-band libraries under src/libraries adhere to that "best practice", the shared framework libraries don't which is partially tracked in https://github.com/dotnet/runtime/issues/43110.
Even though I would love our libraries to get the point where only P2Ps are used so that a dev can directly open VS from a fresh clone, we can't do that yet as msbuild takes too long to evaluate and incrementally build our projects. This will get much better when the [MSBuild server epic](https://github.com/dotnet/msbuild/issues/7315) is implemented. In addition to that, msbuild static graph builds should speed incremental builds up further. On our side we can avoid a ton of time that is currently spent on project evaluation by [letting roslyn generate reference assemblies instead of having dedicated projects](https://github.com/dotnet/runtime/issues/58163). This was quite the verbose response to your statement and probably doesn't  answer what you were asking for but I think it points the spotlight to the actual issue.

Regarding the second desire to being able to build just the reference assemblies, roslyn exposes the `/p:ProduceOnlyReferenceAssembly` property (in addition to the `/p:ProduceReferenceAssembly` switch which is on by default since .NET 5): https://github.com/dotnet/roslyn/blob/c863e7545c964346073d2e0573f30fb7a0f30960/docs/features/refout.md#msbuild. Ideally we would just use that switch whenever we want to just build reference assemblies but we can't do that yet as we don't leverage Roslyn's feature to generate reference assemblies.

To connect both requirements and to finally answer your question, what you actually want is to build the reference assemblies of the shared framework only so that you can enter VS as fast as possible. You don't really care about up-front building the out-of-band reference assemblies as those aren't a requirement to launch VS (again, as those projects well define their dependencies via ProjectReference items). With this change you can achieve the same by invoke the following command from a fresh clone: `build.cmd libs.sfx /p:RefOnly=true`. That is sufficient to afterwards open a VS solution (either via the -vs switch or directly). A `-restore` invocation isn't required anymore since https://github.com/dotnet/runtime/commit/ef5803de62be3c9584a352c1863e3036801731f5. I will make sure that this new switch gets documented.

File: eng/Subsets.props [hunk 1]
```diff
@@ -144,6 +144,9 @@
     <!-- Libs -->
     <SubsetName Include="Libs" Description="The libraries native part, refs and source assemblies, test infra and packages, but NOT the tests (use Libs.Tests to request those explicitly). Equivalent to: $(DefaultLibrariesSubsets)" />
     <SubsetName Include="Libs.Native" Description="The native libraries used in the shared framework." />
+    <SubsetName Include="Libs.Sfx" Description="The managed shared framework libraries." />
+    <SubsetName Include="Libs.Oob" Description="The managed out-of-band libraries." />
+    <!-- TODO: Remove libs.ref and libs.src subsets with the next infra rollout. -->
```
[the remark this answers is not in the document]
Comment (ViktorHofer) [reply eng/Subsets.props:149]: Yes, here's the issue: https://github.com/dotnet/runtime/issues/65492.

**Comment by ViktorHofer** [comment]
As the runtime tests currently depend on sfx+oob reference assemblies (including the shims) being built and I wouldn't want to require them to build the oobs source projects as well, I made sure that the `RefOnly` switch works for the libs.oob subset (+the shims) as well.

Again the PR is ready for review :)

File: eng/NoTargetsSdk.BeforeTargets.targets [hunk 2]
```diff
@@ -1,9 +1,8 @@
 <Project>
 
   <PropertyGroup>
-    <!-- NoTargets SDK needs a TFM set. Use the latest .NETCoreApp TFM that is supported by the SDK.
-         Only set a default if the project isn't multi-targeting. -->
-    <TargetFramework Condition="'$(TargetFramework)' == '' and '$(TargetFrameworks)' == ''">$(NetCoreAppToolCurrent)</TargetFramework>
+    <!-- NoTargets SDK needs a TFM set. Set a default if the project doesn't multi target. -->
+    <TargetFramework Condition="'$(TargetFramework)' == '' and '$(TargetFrameworks)' == ''">$(NetCoreAppCurrent)</TargetFramework>
```
Comment (safern) [inline eng/NoTargetsSdk.BeforeTargets.targets:5]: Didn't we use `NetCoreAppToolCurrent` to make the transition to the new tfm (after the release) smoother or was just a mistake?

File: eng/TraversalSdk.AfterTargets.targets [hunk 3]
```diff
@@ -9,10 +9,17 @@
     <VSTestDependsOn>$(VSTestDependsOn);ResolveP2PReferences</VSTestDependsOn>
     <PackDependsOn>$(PackDependsOn);ResolveP2PReferences</PackDependsOn>
     <PublishDependsOn>$(PublishDependsOn);ResolveP2PReferences</PublishDependsOn>
+    <GetTargetPathDependsOn>ResolveP2PReferences</GetTargetPathDependsOn>
+    <!-- Filter out ProjectReferences which aren't compatible with the project's TargetFramework. -->
+    <OmitIncompatibleProjectReferences>true</OmitIncompatibleProjectReferences>
```
Comment (safern) [inline eng/TraversalSdk.AfterTargets.targets:14]: Why are we adding this?

File: src/libraries/NetCoreAppLibrary.props [hunk 4]
```diff
@@ -1,7 +1,33 @@
...
+      System.Transactions;
+      System.Web;
+      System.Windows;
+      System.Xml;
+      System.Xml.Serialization;
+      System.Xml.Linq;
+      WindowsBase;
+    </NetFxReference>
     <NetCoreAppLibrary>
+      $(NetFxReference)
```
Comment (safern) [inline src/libraries/NetCoreAppLibrary.props:29]: Why do we now need these to appear in the `NetCoreAppLibrary` property?

File: src/libraries/System.Runtime.Experimental/src/System.Runtime.Experimental.proj [hunk 5]
```diff
@@ -1,7 +0,0 @@
-<Project Sdk="Microsoft.Build.Traversal">
```
Comment (safern) [inline src/libraries/System.Runtime.Experimental/src/System.Runtime.Experimental.proj:1]: I guess this was causing a no-op incremental build of the ref?

File: src/libraries/System.Runtime/src/System.Runtime.csproj [hunk 6]
```diff
@@ -20,4 +20,8 @@
     <ProjectReference Include="$(CoreLibProject)" />
     <ProjectReference Include="$(LibrariesProjectRoot)System.Private.Uri\src\System.Private.Uri.csproj" />
   </ItemGroup>
+  <ItemGroup>
+    <!-- Set the ReferenceAssembly metadata to the inbox version of System.Runtime and not to System.Runtime.Experimental. -->
+    <ProjectReference Include="..\ref\System.Runtime.csproj" ReferenceOutputAssembly="false" OutputItemType="ResolvedMatchingContractOverride" />
```
Comment (safern) [inline src/libraries/System.Runtime/src/System.Runtime.csproj:25]: Why is this needed? Is it because of the `ContractProject` override?

File: src/libraries/illink-oob.targets [hunk 7]
```diff
@@ -1,84 +0,0 @@
-<Project>
```
Comment (safern) [inline src/libraries/illink-oob.targets:1]: Where these collapsed into a single targets file?

**Comment by joperezr** [comment]
> With this change you can achieve the same by invoking the following command from a fresh clone: build.cmd libs.sfx /p:RefOnly=true.

While the command you share is simple enough, what is wrong with keeping the libs.ref subset? The whole point of subsets is exactly that: Only build certain parts of the repo which you care about. The other nice thing about subsets is that it is simple enough to figure out all the available subsets as they are right there for you as part of the help of the build.cmd/sh command. Why do we prefer two different mechanisms to filter out projects? (Subsets and properties like RefOnly)

I think one of the main goals we should have with these kinds of changes, is to try to minimize the impact as much as possible to developers existing workflows and by removing subsets (which are likely being used by a lot of devs) we are very likely breaking their workflow. My suggestion here would be to keep the RefOnly logic, but also keep the libs.ref subset, and potentially also keep the libs.src subset which internally would build libs.sfx+libs.oobs subsets.

File: eng/TraversalSdk.AfterTargets.targets (hunk 3 above)
Comment (ViktorHofer) [reply eng/TraversalSdk.AfterTargets.targets:14]: This already existed in the traversal projects: https://github.com/dotnet/runtime/blob/cdc7b457630332c46691ec32c6ba9c0010aa9f7f/src/libraries/src.proj#L9-L11.

The overall reason why this is necessary is already explained in https://github.com/dotnet/arcade/pull/8459. Incompatible target frameworks (based on the BuildTargetFramework property) were filtered out in outer builds. Now that many projects don't have outer builds anymore and use the `TargetFramework` property instead, filtering needs to happen at a different point. Note that this is already implemented in main, this change here just moves it into a shard place as multiple traversal projects depend on filtering.

File: src/libraries/NetCoreAppLibrary.props (hunk 4 above)
Comment (ViktorHofer) [reply src/libraries/NetCoreAppLibrary.props:29]: NetCoreAppLibrary is the list of libraries which compose the shared framework which the .NETFramework shims are part of.

File: src/libraries/System.Runtime.Experimental/src/System.Runtime.Experimental.proj (hunk 5 above)
Comment (ViktorHofer) [reply src/libraries/System.Runtime.Experimental/src/System.Runtime.Experimental.proj:1]: No it wasn't. This project was required in the past when we kicked off packaging via a glob filter: `src/libraries/*/src/*.*proj`. Now that packaging happens implicitly via the `GeneratePackageOnBuild` property, this project isn't required anymore.

File: src/libraries/sfx-projects.proj [hunk 8]
```diff
@@ -0,0 +1,22 @@
+<Project Sdk="Microsoft.Build.Traversal">
```
[the remark this answers is not in the document]
Comment (ViktorHofer) [reply src/libraries/sfx-projects.proj:1]: Today, yes and I will change that, thanks. In the future OOBs won't be split between refs and srcs anymore as we will delete the reference projects as they don't ship anywhere. Today they only exist to enable ApiCompat comparisons to make sure that the API matches between the contract and the implementations and to make it possible to leverage GenAPI to generate the ".ref.cs" file. I hope that we can soon delete at least all the OOB reference projects.

File: src/libraries/illink-oob.targets (hunk 7 above)
Comment (ViktorHofer) [reply src/libraries/illink-oob.targets:1]: The logic of illink-sfx.targets and illink-oob.targets is moved into srx.proj and oob.proj.

File: src/libraries/oob-netcurrent-ref.proj [hunk 9]
```diff
@@ -0,0 +1,18 @@
+<Project Sdk="Microsoft.Build.Traversal">
+  <!-- This project only exists for the /p:RefOnly=true mode.-->
+
+  <PropertyGroup>
+    <TargetFramework>$(NetCoreAppCurrent)</TargetFramework>
+    <!-- Filter ProjectReferences to build the best matching target framework only. -->
+    <FilterTraversalProjectReferences>true</FilterTraversalProjectReferences>
+  </PropertyGroup>
+
+  <!-- Reference all NetCoreAppCurrent out-of-band reference assemblies. -->
```
[the remark this answers is not in the document]
Comment (ViktorHofer) [reply src/libraries/oob-netcurrent-ref.proj]: Quick overview:
- sfx-ref.proj => Builds the NetCoreAppCurrent tfm of all shared framework reference projects
- sfx-src.proj => Builds the NetCoreAppCurrent tfm of all shared framework source projects.
- oob-ref.proj => Builds the NetCoreAppCurrent tfm of all out-of-band reference projects.
- oob-src.proj => Builds the NetCoreAppCurrent tfm of all out-of-band source projects.
- oob-all.proj => Builds either all tfms of all all out-of-band projects when the `-allconfigurations` switch is passed in or just the .NETFramework tfms of all the compatible out-of-band projects when the `-f net48` switch is passed in.

> Why do we need these .proj separation? Why can't we do it on the same .proj file?

That's not possible because we have an ordering requirement when composing the shared framework. The shared framework reference assemblies must be available before the shared framework source projects can be compiled. You can't express such a requirement in a single Traversal project.

**Comment by ViktorHofer** [comment]
@joperezr Thanks again for the feedback. This PR doesn't remove any subsets so I would like to move this discussion over to https://github.com/dotnet/runtime/issues/65492. Also I removed any "// TODO: Remove libs.ref and libs.src" markers in this PR until we reached consensus on if these subsets should be kept or not. Such a change would need to happen as part of an infrastructure batched rollout.

**Comment by ViktorHofer** [comment]
I just resolved merge conflicts. @ericstj who would be best to review and approve the PR now that Santi moved to a different team?

**Comment by jeffhandley** [comment]
Hey @ViktorHofer. We cannot prioritize reviewing this at the moment. It will likely be a week or two before anyone on the team has time. Thanks for your patience and understanding.

File: eng/pipelines/common/templates/runtimes/build-test-job.yml [hunk 10]
```diff
@@ -74,7 +74,7 @@ jobs:
       - ${{ each variable in parameters.variables }}:
         - ${{ variable }}
       - name: liveRuntimeBuildParams
-        value: 'libs.ref -c Release -ci'
+        value: 'libs.sfx+libs.oob /p:RefOnly=true -c Release -ci'
```
Comment (safern) [inline eng/pipelines/common/templates/runtimes/build-test-job.yml:77]: Since there are concerns about removing `libs.ref` and having `/p:RefOnly=true` as the trigger to build the refs only, should we revert this change until a consensus is agreed on the issue open to discuss that?

File: src/tests/Interop/IJW/getRefPackFolderFromArtifacts.ps1 [hunk 11]
```diff
@@ -14,7 +14,7 @@ $refPackPath = "$repoRoot/artifacts/bin/ref/net$majorVersion.$minorVersion"
 
 if (-not (Test-Path $refPackPath))
 {
-    Write-Error "Reference assemblies not found in the artifacts folder at '$refPackPath'. Did you build the libs.ref subset? Did the repo layout change?"
+    Write-Error "Reference assemblies not found in the artifacts folder at '$refPackPath'. Did you invoke 'build.cmd libs.sfx+libs.oob /p:RefOnly=true' to make sure that refs are built? Did the repo layout change?"
```
Comment (safern) [inline src/tests/Interop/IJW/getRefPackFolderFromArtifacts.ps1:17]: Same for this one, should we revert this change as it is not known if libs.ref will be removed, yet?

**Review by safern — APPROVED** [review]
I just did another pass and left a couple of questions, but overall looks good to me as long as everything works fine as expected 😄

File: eng/pipelines/common/templates/runtimes/build-test-job.yml (hunk 10 above)
Comment (ViktorHofer) [reply eng/pipelines/common/templates/runtimes/build-test-job.yml:77]: I would like the infrastructure to use the new subsets but keep the existing ones to not break existing developer workflows.

File: src/tests/Interop/IJW/getRefPackFolderFromArtifacts.ps1 (hunk 11 above)
Comment (ViktorHofer) [reply src/tests/Interop/IJW/getRefPackFolderFromArtifacts.ps1:17]: Same as above.

**Comment by joperezr** [comment]
This broke the build and publishing of the native component as a package of system.io.ports, which is currently blocking the ingestion of runtime in winforms: https://github.com/dotnet/winforms/pull/6767

@carlossanlop and I have a fix for this, we are testing it and will soon send out a PR to fix this.
